File: LeetCode/Python3/Problem 207.py
# LeetCode: Problem 207
# Jonathan Kosasih

#There are a total of n courses you have to take, labeled from 0 to n-1.
#
#Some courses may have prerequisites, for example to take course 0 you have to first take course 1,
#which is expressed as a pair: [0,1]
#
#Given the total number of courses and a list of prerequisite pairs, is it possible for you to finish all courses?
#
#Example 1:
#Input: 2, [[1,0]] 
#Output: true
#Explanation: There are a total of 2 courses to take. 
#             To take course 1 you should have finished course 0. So it is possible.
#             
#Example 2:
#Input: 2, [[1,0],[0,1]]
#Output: false
#Explanation: There are a total of 2 courses to take. 
#             To take course 1 you should have finished course 0, and to take course 0 you should
#             also have finished course 1. So it is impossible.
#------------------------------------------------------------------
import collections
class Solution(object):
    def canFinish(self, numCourses, prerequisites):
        """
        :type numCourses: int
        :type prerequisites: List[List[int]]
        :rtype: bool
        """
        graph = collections.defaultdict(set)
        neighbors = collections.defaultdict(set)
        for course, pre in prerequisites:
            graph[course].add(pre)
            neighbors[pre].add(course)
        queue = [i for i in range(numCourses) if not graph[i]]
        count = 0
        while(queue):
            node = queue.pop()
            count += 1
            for neighbor in neighbors[node]:
                graph[neighbor].remove(node)
                if(not graph[neighbor]):
                    queue.append(neighbor)
        return count == numCourses

sol = Solution()
print(str(sol.canFinish(2, [[1,0]]))) # true
print(str(sol.canFinish(2, [[1,0],[0,1]]))) # false
print(str(sol.canFinish(4, [[2,1], [1,0]]))) # true
print(str(sol.canFinish(3, [[1,0],[1,2],[0,1]]))) # false

File: LeetCode/C++/Problem 207.cpp
// LeetCode: Problem 207
// Jonathan Kosasih
/*
There are a total of numCourses courses you have to take, labeled from 0 to numCourses-1.

Some courses may have prerequisites, for example to take course 0 you have to first take course 1, which is expressed as a pair: [0,1]

Given the total number of courses and a list of prerequisite pairs, is it possible for you to finish all courses?
*/
//-------------------------------------------------------------------

#include <iostream>
#include <vector>
#include <map>
#include <unordered_map>
#include <set>
#include <string>
#include <algorithm> 
#include <stack>
#include <queue>
#include <deque>
#include <numeric>
using std::vector;
using std::map;
using std::unordered_map;
using std::set;
using std::stack;
using std::queue;
using std::deque;
using std::accumulate;

class Solution {
public:
	// STUDY TOPOLOGICAL SORT
	bool canFinish(int numCourses, vector<vector<int>>& prerequisites) {

		vector<unordered_set<int>> matrix(numCourses); // save this directed graph
		for (int i = 0; i < prerequisites.size(); ++i)
			matrix[prerequisites[i][1]].insert(prerequisites[i][0]);

		vector<int> d(numCourses, 0); // in-degree
		for (int i = 0; i < numCourses; ++i)
			for (auto it = matrix[i].begin(); it != matrix[i].end(); ++it)
				++d[*it];

		for (int j = 0, i; j < numCourses; ++j)
		{
			for (i = 0; i < numCourses && d[i] != 0; ++i); // find a node whose in-degree is 0

			if (i == numCourses) // if not find
				return false;

			d[i] = -1;
			for (auto it = matrix[i].begin(); it != matrix[i].end(); ++it)
				--d[*it];
		}

		return true;
	}
};
int main()
{
    //std::cout << "Hello World!\n";
}